File: VideoStreamingBackend/src/domain/entities/ViewerSession.ts
/**
 * Domain entity: Viewer session tracks a single WebRTC consumer connection.
 * Each viewer has a transport, a video consumer, and optionally an audio consumer.
 */

export interface ViewerSession {
  /** Unique viewer session ID */
  id: string;
  /** Camera being watched */
  cameraId: string;
  /** Authenticated user ID */
  userId: string;
  /** mediasoup WebRtcTransport ID */
  transportId: string;
  /** mediasoup video Consumer ID */
  videoConsumerId: string | null;
  /** mediasoup audio Consumer ID */
  audioConsumerId: string | null;
  /** Connection state */
  state: ViewerState;
  /** When the viewer connected */
  connectedAt: Date;
  /** Last ICE state update */
  lastActiveAt: Date;
  /** Client user-agent (for analytics) */
  userAgent: string | null;
}

export type ViewerState =
  | 'connecting'    // Transport created, ICE not complete
  | 'connected'     // DTLS established, consuming media
  | 'paused'        // Consumer paused (tab background, etc.)
  | 'disconnected'; // Transport closed

File: StreamingBackend/internal/domain/entities/viewer_session.go
package entities

import "time"

// ViewerState represents a WebRTC viewer's connection state.
type ViewerState string

const (
	ViewerStateConnecting   ViewerState = "connecting"
	ViewerStateConnected    ViewerState = "connected"
	ViewerStatePaused       ViewerState = "paused"
	ViewerStateDisconnected ViewerState = "disconnected"
)

// ViewerSession tracks a single WebRTC peer connection for a viewer.
type ViewerSession struct {
	ID           string      `json:"id"`
	CameraID     string      `json:"camera_id"`
	UserID       string      `json:"user_id"`
	State        ViewerState `json:"state"`
	ConnectedAt  time.Time   `json:"connected_at"`
	LastActiveAt time.Time   `json:"last_active_at"`
	UserAgent    string      `json:"user_agent"`
}

// NewViewerSession creates a new viewer session.
func NewViewerSession(id, cameraID, userID, userAgent string) *ViewerSession {
	now := time.Now().UTC()
	return &ViewerSession{
		ID:           id,
		CameraID:     cameraID,
		UserID:       userID,
		State:        ViewerStateConnecting,
		ConnectedAt:  now,
		LastActiveAt: now,
		UserAgent:    userAgent,
	}
}
